packages/flutter/lib/src/material/carousel.dart

```diff
@@ -730,13 +1426,27 @@ class CarouselController extends ScrollController {
   @override
   ScrollPosition createScrollPosition(ScrollPhysics physics, ScrollContext context, ScrollPosition? oldPosition) {
     assert(_carouselState != null);
-    final double itemExtent = _carouselState!._itemExtent;
+    final List<int>? weights = _carouselState!._weights;
+    final double? itemExtent = _carouselState!._itemExtent;
+    int expandedItem = initialItem;
+
+    if (weights != null && !_carouselState!._consumeMaxWeight) {
```

Comment:
Related to the above question: What if the values read from _carouselState change after they have been read? Do we need to inform the controller and redo some of these calculations? Would also be good to add a test for that case.

Reply:
We've been chatting about this. I think what is needed here is to update the position, not the controller. The values provided to the controller are just passed to (and only used by) the position, which calculates what item is at what pixel, or what pixel is at which item based on those values.
So if they change here, we should update `_controller.position.VALUE` so it can make the right calculations based on the new info.

Converting the properties in the carousel scroll position to setters/getters will make this possible, with an update to the pixels when the value changes. (See _PagePosition.viewportFraction). And thanks Qun for mentioning PageView, looking at that really helped!

What do you all think?

Reply:
This makes sense to me:) Trying to add setters and getters for `itemExtent` and `flexWeights` in `_CarouselPosition` and setting they to new values when they are updated. Thanks for the suggestions!

Reply:
Just updated! `_flexWeights` and `_consumeMaxWeight` are changed to getters in `_CarouselViewState` class. `_itemExtent` should be constrained to be smaller than the viewport in the implementation, so I keep it as is. 

And when we update `flexWeights`, the layout will change to match the new weight list but the item that occupies the max weight stays the same. When we update `itemExtent`, the item main-axis extent will change but leading item also stays the same.

<details>
<summary>Demo 1 (Updating weights respects `initialItem`)</summary>

https://github.com/flutter/flutter/assets/36861262/f8b00fd0-da7b-47f0-b194-67bb45a13d49

</details>

<details>
<summary>Demo 2 (Updating weights respects the item that currently occupied the max weight)</summary>

https://github.com/flutter/flutter/assets/36861262/efdff60a-8a12-491b-a3e7-f69fd1e83177

</details>